mlir/lib/Conversion/TosaToMIOpen/TosaToMIOpen.cpp

```diff
@@ -215,15 +233,52 @@ class ConvConverter final : public OpConversionPattern<tosa::Conv2DOp> {
           {getAffineDimExpr(0, context), getAffineDimExpr(1, context),
            getAffineDimExpr(2, context), getAffineDimExpr(3, context)});
 
-      auto bias_expand_mr = rewriter.create<memref::ExpandShapeOp>(
+      auto bias_expand_mr = rw.create<memref::ExpandShapeOp>(
           loc, newType, bias_mr, reassociations);
 
-      auto bias_t = rewriter.create<bufferization::ToTensorOp>(loc, bias_expand_mr);
-      output = rewriter.create<tosa::AddOp>(loc, op.getType(),
-                                            ValueRange{conv_output_t, bias_t});
+      auto bias_t = rw.create<bufferization::ToTensorOp>(loc, bias_expand_mr);
+      output = rw.create<tosa::AddOp>(loc, op.getType(),
+                                      ValueRange{conv_output_t, bias_t});
+    }
+    return success();
+  }
+};
+
```

Comment:
We may want to have miopen.matmul because of the batch support, which might be possible via using group as you mentioned. But group dimension goes into the C dim and requires transpose at the end. Maybe not if miopen.conv support ANY layout.

Reply:
NHWC is a better layout since the inner dimension corresponds to the K dimension of GEMM and we can use coalesce loads. But would be good to transpose input to also NHWC.

When we get general transpose support this will be a good optimization.

Reply:
I think, coalesced load can be enabled only when the original data layout is in MxK and NxK, i.e. matB is transposed. Otherwise HW can't coalesce the vertical vector load.
I just prefer to split compilation on tensors and compilation on buffers but this is still fine.
However in the future, I think we will need low level operation miopen.matmul with transpose option to each matrix, like BLAS GEMM API has (+more layout if needed) .

Reply:
Yeah, allowing for transpose and adding a `miopen.matmul` is a good call

One thing I'll note is that NCHW is faster than NHWC from what I've seen

Reply:
@krzysz00 is that true for 1x1. It is already a matrix.

Reply:
I'll double-check perf results later